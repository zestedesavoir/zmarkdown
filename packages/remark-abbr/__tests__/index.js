import dedent from 'dedent'
import { unified } from 'unified'
import reParse from 'remark-parse'
import stringify from 'rehype-stringify'
import remark2rehype from 'remark-rehype'
import remarkStringify from 'remark-stringify'

import remarkAbbr from '../lib/index'

const render = (text, config) => {
  const result = unified()
    .use(reParse)
    .use(remarkAbbr, config)
    .use(remark2rehype, {
      handlers: {
        abbrDefinition: () => undefined,
      }
    })
    .use(stringify)
    .processSync(text)
  return String(result)
}

const renderToMarkdown = (text, config) => {
  const result = unified()
    .use(reParse)
    .use(remarkAbbr, config)
    .use(remarkStringify)
    .processSync(text)

  return String(result)
}

const configToTest = {
  'no-config': undefined,
  'empty object': {},
  expandFirst: {expandFirst: true},
}

for (const [configName, config] of Object.entries(configToTest)) {
  it(`${configName} renders references`, () => {
    const contents = render(dedent`
      This is an abbreviation: REF.
      ref and REFERENCE should be ignored.

      Here is another one in a link: [FOO](http://example.com).

      Here is the first one in a link: [REF](http://example.com).

      *[REF]: Reference
      *[FOO]: Reference
    `, config)

    expect(contents).toMatchSnapshot()
  })


  it(`${configName} passes the first regression test`, () => {
    const contents = render(dedent`
      The HTML specification is maintained by the W3C:\
      [link](https://w3c.github.io/html/), this line had an abbr before link.

      A line with [a link](http://example.com) before an abbr: HTML.

      *[HTML]: Hyper Text Markup Language
      *[W3C]:  World Wide Web Consortium
    `, config)

    expect(contents).toMatchSnapshot()
  })

  it(`${configName} passes the second regression test`, () => {
    const contents = render(dedent`
      The HTML specification is maintained by the W3C:\
      [link](https://w3c.github.io/html/), this line had an abbr before **link** HTML.

      A line with [a link](http://example.com) before an abbr: HTML.

      *[HTML]: Hyper Text Markup Language
      *[W3C]:  World Wide Web Consortium
    `, config)

    expect(contents).toMatchSnapshot()
  })

  it(`${configName} passes the retro test`, () => {
    const input = dedent`
      An ABBR: "REF", ref and REFERENCE should be ignored.

      The HTML specification is maintained by the W3C.

      *[REF]: Reference
      *[ABBR]: This gets overridden by the next one.
      *[ABBR]: Abbreviation
      *[HTML]: Hyper Text Markup Language
      *[W3C]:  World Wide Web Consortium
    `

    const html = render(input)
    expect(html).toMatchSnapshot()

    const markdown = renderToMarkdown(input)
    expect(markdown).toMatchSnapshot()
  })

  it(`${configName} no reference`, () => {
    const contents = render(dedent`
      No reference!
    `, config)

    expect(contents).toMatchSnapshot()
  })

  test('compiles to markdown', () => {
    const md = dedent`
      *abbr* HTML

      > HTML inside quote

      *[abbr]: abbreviation
      *[noabbr]: explanation that does not match
      *[HTML]: HyperText Markup Language
    `
    const contents = renderToMarkdown(md)
    expect(contents).toMatchSnapshot()

    const contents1 = renderToMarkdown(md)
    const contents2 = renderToMarkdown(contents1)

    expect(contents1).toBe(contents2)
  })

  it(`${configName} handles abbreviations ending with a period`, () => {
    const contents = render(dedent`
      A.B.C. and C-D%F. foo

      *[A.B.C.]: ref1
      *[C-D%F.]: ref2
    `, config)

    expect(contents).toContain(`<abbr title="ref1">A.B.C.</abbr>`)
    expect(contents).toContain(`<abbr title="ref2">C-D%F.</abbr>`)
  })

  it(`${configName} does not parse words starting with abbr`, () => {
    const contents = render(dedent`
      ABC ABC ABC

      *[AB]: ref1
    `, config)

    expect(contents).not.toContain('<abbr')
  })

  it(`${configName} does not parse words ending with abbr`, () => {
    const contents = render(dedent`
      ABC ABC ABC

      *[BC]: ref1
    `, config)

    expect(contents).not.toContain('<abbr')
  })

  it(`${configName} does not parse words containing abbr`, () => {
    const contents = render(dedent`
      ABC ABC ABC

      *[B]: ref1
    `, config)

    expect(contents).not.toContain('<abbr')
  })

  it(`${configName} does not break with references in their own paragraphs`, () => {
    const contents = render(dedent`
      Here is a test featuring abc and def

      *[abc]: A B C

      *[def]: D E F
    `, config)

    expect(contents).toMatchSnapshot()
  })
}
