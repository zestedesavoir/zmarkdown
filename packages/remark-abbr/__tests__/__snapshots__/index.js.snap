// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiles to markdown 1`] = `
"_abbr_ HTML

> HTML inside quote

*[abbr]: abbreviation
*[HTML]: HyperText Markup Language"
`;

exports[`compiles to markdown 2`] = `
"_abbr_ HTML

> HTML inside quote

*[abbr]: abbreviation
*[HTML]: HyperText Markup Language"
`;

exports[`compiles to markdown 3`] = `
"_abbr_ HTML

> HTML inside quote

*[abbr]: abbreviation
*[HTML]: HyperText Markup Language"
`;

exports[`empty object does not break with references in their own paragraphs 1`] = `"<p>Here is a test featuring <abbr title="A B C">abc</abbr> and <abbr title="D E F">def</abbr></p>"`;

exports[`empty object no reference 1`] = `"<p>No reference!</p>"`;

exports[`empty object passes the first regression test 1`] = `
"<p>The <abbr title="Hyper Text Markup Language">HTML</abbr> specification is maintained by the <abbr title="World Wide Web Consortium">W3C</abbr>:<a href="https://w3c.github.io/html/">link</a>, this line had an abbr before link.</p>
<p>A line with <a href="http://example.com">a link</a> before an abbr: <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>"
`;

exports[`empty object passes the retro test 1`] = `
"<p>An <abbr title="Abbreviation">ABBR</abbr>: "<abbr title="Reference">REF</abbr>", ref and REFERENCE should be ignored.</p>
<p>The <abbr title="Hyper Text Markup Language">HTML</abbr> specification is maintained by the <abbr title="World Wide Web Consortium">W3C</abbr>.</p>"
`;

exports[`empty object passes the retro test 2`] = `
"An ABBR: "REF", ref and REFERENCE should be ignored.

The HTML specification is maintained by the W3C.

*[ABBR]: Abbreviation
*[REF]: Reference
*[HTML]: Hyper Text Markup Language
*[W3C]: World Wide Web Consortium"
`;

exports[`empty object passes the second regression test 1`] = `
"<p>The <abbr title="Hyper Text Markup Language">HTML</abbr> specification is maintained by the <abbr title="World Wide Web Consortium">W3C</abbr>:<a href="https://w3c.github.io/html/">link</a>, this line had an abbr before <strong>link</strong> <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>
<p>A line with <a href="http://example.com">a link</a> before an abbr: <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>"
`;

exports[`empty object renders references 1`] = `
"<p>This is an abbreviation: <abbr title="Reference">REF</abbr>.
ref and REFERENCE should be ignored.</p>
<p>Here is another one in a link: <a href="http://example.com"><abbr title="Reference">FOO</abbr></a>.</p>
<p>Here is the first one in a link: <a href="http://example.com"><abbr title="Reference">REF</abbr></a>.</p>"
`;

exports[`expandFirst does not break with references in their own paragraphs 1`] = `"<p>Here is a test featuring A B C (<abbr title="A B C">abc</abbr>) and D E F (<abbr title="D E F">def</abbr>)</p>"`;

exports[`expandFirst no reference 1`] = `"<p>No reference!</p>"`;

exports[`expandFirst passes the first regression test 1`] = `
"<p>The Hyper Text Markup Language (<abbr title="Hyper Text Markup Language">HTML</abbr>) specification is maintained by the World Wide Web Consortium (<abbr title="World Wide Web Consortium">W3C</abbr>):<a href="https://w3c.github.io/html/">link</a>, this line had an abbr before link.</p>
<p>A line with <a href="http://example.com">a link</a> before an abbr: <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>"
`;

exports[`expandFirst passes the retro test 1`] = `
"<p>An <abbr title="Abbreviation">ABBR</abbr>: "<abbr title="Reference">REF</abbr>", ref and REFERENCE should be ignored.</p>
<p>The <abbr title="Hyper Text Markup Language">HTML</abbr> specification is maintained by the <abbr title="World Wide Web Consortium">W3C</abbr>.</p>"
`;

exports[`expandFirst passes the retro test 2`] = `
"An ABBR: "REF", ref and REFERENCE should be ignored.

The HTML specification is maintained by the W3C.

*[ABBR]: Abbreviation
*[REF]: Reference
*[HTML]: Hyper Text Markup Language
*[W3C]: World Wide Web Consortium"
`;

exports[`expandFirst passes the second regression test 1`] = `
"<p>The Hyper Text Markup Language (<abbr title="Hyper Text Markup Language">HTML</abbr>) specification is maintained by the World Wide Web Consortium (<abbr title="World Wide Web Consortium">W3C</abbr>):<a href="https://w3c.github.io/html/">link</a>, this line had an abbr before <strong>link</strong> <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>
<p>A line with <a href="http://example.com">a link</a> before an abbr: <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>"
`;

exports[`expandFirst renders references 1`] = `
"<p>This is an abbreviation: Reference (<abbr title="Reference">REF</abbr>).
ref and REFERENCE should be ignored.</p>
<p>Here is another one in a link: <a href="http://example.com">Reference (<abbr title="Reference">FOO</abbr>)</a>.</p>
<p>Here is the first one in a link: <a href="http://example.com"><abbr title="Reference">REF</abbr></a>.</p>"
`;

exports[`no-config does not break with references in their own paragraphs 1`] = `"<p>Here is a test featuring <abbr title="A B C">abc</abbr> and <abbr title="D E F">def</abbr></p>"`;

exports[`no-config no reference 1`] = `"<p>No reference!</p>"`;

exports[`no-config passes the first regression test 1`] = `
"<p>The <abbr title="Hyper Text Markup Language">HTML</abbr> specification is maintained by the <abbr title="World Wide Web Consortium">W3C</abbr>:<a href="https://w3c.github.io/html/">link</a>, this line had an abbr before link.</p>
<p>A line with <a href="http://example.com">a link</a> before an abbr: <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>"
`;

exports[`no-config passes the retro test 1`] = `
"<p>An <abbr title="Abbreviation">ABBR</abbr>: "<abbr title="Reference">REF</abbr>", ref and REFERENCE should be ignored.</p>
<p>The <abbr title="Hyper Text Markup Language">HTML</abbr> specification is maintained by the <abbr title="World Wide Web Consortium">W3C</abbr>.</p>"
`;

exports[`no-config passes the retro test 2`] = `
"An ABBR: "REF", ref and REFERENCE should be ignored.

The HTML specification is maintained by the W3C.

*[ABBR]: Abbreviation
*[REF]: Reference
*[HTML]: Hyper Text Markup Language
*[W3C]: World Wide Web Consortium"
`;

exports[`no-config passes the second regression test 1`] = `
"<p>The <abbr title="Hyper Text Markup Language">HTML</abbr> specification is maintained by the <abbr title="World Wide Web Consortium">W3C</abbr>:<a href="https://w3c.github.io/html/">link</a>, this line had an abbr before <strong>link</strong> <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>
<p>A line with <a href="http://example.com">a link</a> before an abbr: <abbr title="Hyper Text Markup Language">HTML</abbr>.</p>"
`;

exports[`no-config renders references 1`] = `
"<p>This is an abbreviation: <abbr title="Reference">REF</abbr>.
ref and REFERENCE should be ignored.</p>
<p>Here is another one in a link: <a href="http://example.com"><abbr title="Reference">FOO</abbr></a>.</p>
<p>Here is the first one in a link: <a href="http://example.com"><abbr title="Reference">REF</abbr></a>.</p>"
`;
